hedera-node/hedera-app/src/main/java/com/hedera/node/app/uploader/MinioBucketUploader.java

```diff
@@ -0,0 +1,206 @@
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+@Singleton
+public class MinioBucketUploader implements CloudBucketUploader {
+    private static final Logger logger = LogManager.getLogger(MinioBucketUploader.class);
+    private final List<MinioClient> minioClients;
+    private final String bucketName;
+    private final BucketProvider provider;
+    private final ExecutorService uploadExecutor;
+    private final int maxRetryAttempts;
+
+    public class MinioClientFactory {
+        public static List<MinioClient> createClients(List<CompleteBucketConfig> configs) {
+            List<MinioClient> clients = new ArrayList<>();
+            for (CompleteBucketConfig config : configs) {
+                MinioClient.Builder builder = MinioClient.builder()
+                        .endpoint(config.endpoint())
+                        .credentials(
+                                config.credentials().accessKey(),
+                                Arrays.toString(config.credentials().secretKey()));
+                // Add region only if the provider is AWS
+                if (config.provider().equals(BucketProvider.AWS) && config.region() != null) {
+                    builder.region(config.region());
+                }
+                // Build and add the client to the list
+                clients.add(builder.build());
+            }
+            return clients;
+        }
+    }
+
+    @Inject
+    public MinioBucketUploader(
+            BucketConfigurationManager bucketConfigurationManager,
+            @CommonExecutor ExecutorService executor,
+            ConfigProvider configProvider) {
+        List<CompleteBucketConfig> completeBucketConfigs = bucketConfigurationManager.getCompleteBucketConfigs();
+        this.uploadExecutor = executor;
+        this.bucketName = completeBucketConfigs.getFirst().bucketName();
+        this.provider = BucketProvider.valueOf(completeBucketConfigs.getFirst().provider());
+        this.maxRetryAttempts = configProvider
+                .getConfiguration()
+                .getConfigData(BlockStreamConfig.class)
+                .uploadRetryAttempts();
+        this.minioClients = MinioClientFactory.createClients(bucketConfigurationManager.getCompleteBucketConfigs());
+    }
+
+    private String calculateMD5Hash(Path filePath) throws IOException, NoSuchAlgorithmException {
+        // Calculate MD5 hash
+        byte[] fileBytes = Files.readAllBytes(filePath); // Read the file content as bytes
+        MessageDigest md = MessageDigest.getInstance("MD5");
```

Comment:
Perhaps we can have the `MessageDigest` instance created once for a `MinioBucketUploader`?